### PR TITLE
Cxops 2866

### PR DESCRIPTION
**Pull request by vfangellation**
## Ready State
**Ready**

## Synopsis
First draft of unittest for ef_open starting with the ef_utils module.
[CXOPS-2866](https://ellation.atlassian.net/browse/CXOPS-2866)

## Changelog
Added unit tests for all the methods in ef_utils.py

Also moved all the tests from src into the test folder, so you'll see some changes to the imports for the other test files.

## General Comment On Request
This is again a first draft, so anything about standards, comment blocks, patterns, in depth testing, too much testing, too much mocking, not enough, different types of inputs, unit vs integration tests comments or questions are welcome.

It's also come to my attention that the setup we have for the dummy ef_site_config.py located in tests/test_config/ might need to be changed to be more variable dependent. Meaning instead of the tests relying on hard coded values matching the expected answer in ef_site_config.py I should reference the variables directly so that other people who want to test their config can drop their own version of ef_site_config.py into the folder. Example would be the different environment names. Would welcome comments on this.

07/07/2017 Update: Discussing with various team members, the definition of unit testing has narrowed down the scope on what should be tested and mocked, and what should be left to integration testing due to the nature of how failure cases can change or be incorrectly assumed in unit testing scenarios. Thus, many functions that overly mocked the 3rd party libraries for failure cases have been removed. Coverage has gone from 100% to 88% now.

07/11/2017 Update: Just modified the directory structure inside tests to now have unit_tests so we can separate the integration_tests in another folder in the future.

## Testing
Testing requires the following modules be installed in pip
coverage
pytest
mock

along with whatever modules are needed for the project itself. Will be writing a confluence page with all the details on the environment.

Note that two tests fail due to bugs found in the code. http_get_metadata does not wrap the status code inside str() when a failure occurs in the error message so an exception is raised there.

pull_repo() does not handle https:// credentials correctly and fails on that one.
```
(venv-2.7) vfang-mlt:ef-open vfang$ pytest tests/unit_tests/test_ef_utils.py --cov ef_utils -v --cov-report term-missing
==================================================== test session starts ====================================================
platform darwin -- Python 2.7.10, pytest-3.1.2, py-1.4.34, pluggy-0.4.0 -- /Users/vfang/venv-2.7/bin/python
cachedir: .cache
rootdir: /Users/vfang/workspace/ef-open, inifile:
plugins: cov-2.5.1
collected 29 items 

tests/unit_tests/test_ef_utils.py::TestEFUtils::test_create_aws_clients PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_create_aws_clients_no_profile PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_env_valid PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_env_valid_invalid_envs PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_fail_with_None_message PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_fail_with_empty_string PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_fail_with_message PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_fail_with_message_and_exception_data PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_account_alias PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_account_alias_invalid_env PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_env_short PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_env_short_invalid_envs PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_instance_aws_context PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_instance_aws_context_ec2_invalid_environment_exception PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_get_instance_aws_context_metadata_exception PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_global_env_valid PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_global_env_valid_non_scoped_envs PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_instance_env PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_instance_env_exception PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_instance_role PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_instance_role_exception PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_metadata_200_status_code PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_http_get_metadata_non_200_status_code FAILED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_pull_repo_https_credentials FAILED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_pull_repo_incorrect_branch PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_pull_repo_incorrect_repo PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_pull_repo_ssh_credentials PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_whereami_local PASSED
tests/unit_tests/test_ef_utils.py::TestEFUtils::test_whereami_unknown PASSED

---------- coverage: platform darwin, python 2.7.10-final-0 ----------
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
src/ef_utils.py     130     15    88%   83-84, 89-90, 135-136, 157-158, 164-165, 170-171, 198-199, 235


========================================================= FAILURES ==========================================================
__________________________________ TestEFUtils.test_http_get_metadata_non_200_status_code ___________________________________

self = <tests.unit_tests.test_ef_utils.TestEFUtils testMethod=test_http_get_metadata_non_200_status_code>
mock_urllib2 = <MagicMock name='urlopen' id='4490506832'>

    @patch('urllib2.urlopen')
    def test_http_get_metadata_non_200_status_code(self, mock_urllib2):
      """
        Test http_get_metadata to retrieve ami-id and get a non 200 status code.
    
        Args:
          mock_urllib2:  MagicMock, returns back a non 200 status code.
    
        Returns:
          None
    
        Raises:
          AssertionError if any of the assert checks fail
        """
      mock_response = Mock(name="Always non-200 Status Code")
      mock_response.getcode.return_value = 400
      mock_urllib2.return_value = mock_response
      with self.assertRaises(IOError) as exception:
>       http_get_metadata("ami-id")

tests/unit_tests/test_ef_utils.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

metadata_path = 'http://169.254.169.254/latest/meta-data/ami-id', timeout = 5

    def http_get_metadata(metadata_path, timeout=__HTTP_DEFAULT_TIMEOUT_SEC):
      """
      Fetch AWS metadata from http://169.254.169.254/latest/meta-data/<metadata_path>
      ARGS:
        metadata_path - the optional path and required key to the EC2 metadata (e.g. "instance-id")
      RETURN:
        response content on success
      RAISE:
        URLError if there was a problem reading metadata
      """
      metadata_path = __METADATA_PREFIX + metadata_path
      try:
        response = urllib2.urlopen(metadata_path, None, timeout)
        if response.getcode() != 200:
>         raise IOError("Non-200 response " + response.getcode() + " reading " + metadata_path)
E         TypeError: cannot concatenate 'str' and 'int' objects

src/ef_utils.py:66: TypeError
_______________________________________ TestEFUtils.test_pull_repo_https_credentials ________________________________________

self = <tests.unit_tests.test_ef_utils.TestEFUtils testMethod=test_pull_repo_https_credentials>
mock_check_output = <MagicMock name='check_output' id='4490638672'>

    @patch('subprocess.check_output')
    def test_pull_repo_https_credentials(self, mock_check_output):
      """
        Tests the pull_repo by mocking the subprocess.check_output to return git http credentials.
    
        Args:
          mock_check_output: MagicMock, returns valid git responses in order of being called, with the
          repo coming from the ef_site_config.py
    
        Returns:
          None
    
        Raises:
          AssertionError if any of the assert checks fail
        """
      mock_check_output.side_effect = [
        "origin\thttps://user@" + EFSiteConfig.EF_REPO + ".git",
        EFSiteConfig.EF_REPO_BRANCH
      ]
      try:
        pull_repo()
      except RuntimeError as exception:
>       self.fail("Exception occurred during test_pull_repo_ssh_credentials: " + exception.message)
E       AssertionError: Exception occurred during test_pull_repo_ssh_credentials: Must be in github.com/company/fake_repo repo. Current repo is: user

tests/unit_tests/test_ef_utils.py:414: AssertionError
============================================ 2 failed, 27 passed in 2.52 seconds ============================================

```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@
 .pants.workdir.file_lock
 .pids
 dist
+.coverage
```
Comment (bobsut) [inline .gitignore]: needs a terminating newline

Comment (vfangellation) [inline .gitignore]: Added

**Comment by theJWong** [comment]
Let's add some documentation in the `docs/` directory in regards to these unit-test.

File: pylintrc [hunk 2]
```diff
@@ -186,7 +186,7 @@ max-nested-blocks=5
 [FORMAT]
 
 # Maximum number of characters on a single line.
-max-line-length=100
+max-line-length=120
```
Comment (vfangellation) [inline pylintrc]: The team of engineers in deep engagement have agreed to a compromise of 100.

**Comment by vfangellation** [comment]
Will make documentation available in next PR, @theJWong

**Pull request merged**